core/src/test/java/org/elasticsearch/common/xcontent/ObjectParserTests.java

```diff
@@ -593,6 +596,131 @@ public void testIgnoreUnknownArrays() throws IOException {
         assertEquals(s.test, "foo");
     }
 
+    /**
+     * test parsing fields with a random name
+     */
+    public void testUnknownFieldParser() throws IOException {
+        XContentBuilder b = XContentBuilder.builder(XContentType.JSON.xContent());
+        String randomName = "unknown#" + randomIntBetween(1, 9);
+        b.startObject();
+        {
+            b.field("test", "foo");
+            b.field("shouldBeIgnored", "foo3");
+            b.field(randomName, "foo2");
+        }
+        b.endObject();
+        b = shuffleXContent(b);
+        XContentParser parser = createParser(JsonXContent.jsonXContent, b.bytes());
+
+        class TestObject {
+            public String test;
+            public String unknown;
+            public String name;
+
+            public void setUnknown(String name, String unknown) {
+                this.name = name;
+                this.unknown = unknown;
+            }
+        }
+
+        boolean ignoreUnknown = randomBoolean();
+        ObjectParser<TestObject, Void> objectParser = new ObjectParser<>("testParser", ignoreUnknown, null);
+        objectParser.declareField((i, c, x) -> c.test = i.text(), new ParseField("test"), ObjectParser.ValueType.STRING);
+        objectParser.declareMatchFieldParser((value, tuple) -> {
+            value.setUnknown(tuple.v1(), tuple.v2());
+        }, (p, c) -> {
+            String name = p.currentName();
+            String value = p.text();
+            return new Tuple<>(name, value);
+        }, s -> s.contains("#"), ObjectParser.ValueType.STRING);
+
+        if (ignoreUnknown == true) {
```

Comment:
No, I might have added that for readablility but I don't like it now that I see it.